src/FreeAge/client/game_controller.cpp

```diff
@@ -394,6 +397,54 @@ void GameController::HandleUnitMovementMessage(const QByteArray& data) {
   unit->SetMovementSegment(currentGameStepServerTime, startPoint, speed, action, map.get(), match.get());
 }
 
+void GameController::HandleUnitGarrisonMessage(const QByteArray& data) {
+  if (data.size() < 4) {
+    LOG(ERROR) << "Received a too short UnitGarrison message";
+    return;
+  }
+  const char* buffer = data.data();
+  
+  u32 unitId = mango::uload32(buffer + 0);
+  u32 targetObjectId = mango::uload32(buffer + 4);
+
+  auto unitIt = map->GetObjects().find(unitId);
+  if (unitIt == map->GetObjects().end()) {
+    LOG(ERROR) << "Received a UnitGarrison message for an object ID that is not in the map.";
+    return;
+  }
+  if (!unitIt->second->isUnit()) {
+    LOG(ERROR) << "Received a UnitGarrison message for an object ID that is a different type than a unit.";
+    return;
+  }
+  auto targetObjectIt = map->GetObjects().find(targetObjectId);
+  if (targetObjectIt == map->GetObjects().end()) {
+    LOG(ERROR) << "Received a UnitGarrison message with target object ID that is not in the map.";
+    return;
+  }
+
+  ClientUnit* unit = AsUnit(unitIt->second);
+  GarrisonUnit(unit, targetObjectIt->second, !unit->IsGarrisoned());
+}
+
+void GameController::GarrisonUnit(ClientUnit* unit, ClientObject* targetObject, bool enter) {
```

Comment:
> do not share any functionality

They will when gamelogic is implemented. I would prefer to have the gamelogic of a relic garrison and garrisoning in the same place inside the `GarrisonUnit` instead of having spread apart. In any case the name should probably change. I will leave this for you or whoever implements some gamelogic there...

Plus: There only should be called once (from the server message)

Reply:
Could be refactor to similar to the server side functions (see last comment)

Reply:
Renamed to `ChangeUnitGarrisonStatus`